### PR TITLE
Optimize performance in table upsert backwards compatibility scenarios

### PR DESCRIPTION
**Pull request by GiaJordan**
Add functionality to the `synapseStorage` object to make calls to the synapse REST API by way of the functionality exposed in the `synapsePythonClient`.

The method for making the calls implemented in the `synapseStorage` object is intended for general use, but due to current deadlines and time constraints is not fully generalized and is mostly relevant to the two calls that are necessary for renaming table columns when necessary during upsert operations. Also, while async calls are made, they are not necessarily used asynchronously on the schematic side, again due to time constraints, so future development should be done on these fronts to improve performance and usability.

Also changed the structure of the `TableOperations` class since it now has more helper methods and allowing instances to reference themselves simplifies things.

### CODE REVIEW

**Comment by GiaJordan** [comment]
@linglp the code for the API calls will be in this PR

**Comment by GiaJordan** [comment]
Testing results with new method:

<table>
<thead>
<tr>
<th>Performance Time (seconds)</th>
<th>CLI using python client ops</th>
<th>CLI using synapse API calls</th>
<th>API using python client ops</th>
<th>API using synapse API calls</th>
</thead>
<tbody>
<tr>
<td>Adding 200 rows to a table with ~1 existing row</td>
<td>47</td>
<td>43</td>
<td>116</td>
<td>42</td>
</tr>
<tr>
<td>Adding 200 rows to a table with ~200 existing rows</td>
<td>41</td>
<td>43</td>
<td>148</td>
<td>48</td>
</tr>
</tbody>


@milen-sage @AmyHeiser

File: schematic/store/synapse.py
Comment (GiaJordan) [inline schematic/store/synapse.py:2163]: @thomasyu888 here's the relevant chunk of code for renaming a column in an existing table

File: schematic/store/synapse.py
Comment (thomasyu888) [inline schematic/store/synapse.py:299]: You might be aware of this, but the synapse client does have these:

```
import synapseclient
syn = synapseclient.login()
syn.restPOST
syn.restGET
syn.restPUT
syn.restDELETE
```

Those are heavily tested, so it might be nice to use those if it captures everything you need.

File: schematic/store/synapse.py
Comment (thomasyu888) [inline schematic/store/synapse.py:2289]: Nit: This can be replaced with

```
self.synStore.syn.restPOST("column", body=columnModelDict)
```

File: schematic/store/synapse.py
Comment (thomasyu888) [inline schematic/store/synapse.py:2310]: Same thing here about using existing client `syn.restPOST`.  I do like the `send_api_request` function though!

File: tests/test_store.py
Comment (thomasyu888) [inline tests/test_store.py:221]: Using existing client functionality will allow you to remove this test.  That said, i think what's more important is to test `_update_table_uuid_column` this function.

File: schematic/store/synapse.py
Comment (thomasyu888) [inline schematic/store/synapse.py:2289]: This can be done with the client:

```
new_col = synapseclient.Column(name='Id', columnType='STRING', maximumSize=64, maximumListLength=1)
self.synStore.syn.store(new_col)
```

**Review by thomasyu888 — COMMENTED** [review]
nice work @GiaJordan ! Just some minor comments that might help with maintenance of the code by offloading some of that to the Python client.

Overall, I think FAIR should decide on a standard: snake case or camelCase and stick with it for consistent code readability.

File: schematic/store/synapse.py
Comment (thomasyu888) [inline schematic/store/synapse.py:2105]: Nit: `self.tableName` is default value is None and its type is `str`. I recommend to be specific

```
if self.tableName is not None
```

Not that this will happen, but if you set the value as `0`. you will get something like this:

```
tableName = 0
datasetName = "foo"
if not tableName:
    tableName = datasetName + 'table'
print(tableName)
# footable
```

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py:2163]: @thomasyu888 for sure, hope you find it useful
For this chuck of code, right now I don't plan for this to be reused. When speaking with @andrewelamb he mentioned that this isn't something we want to be doing often, in the context of treating Synapse tables like a database. The preferred method would be for users to rebuild their schema and reupload if they need to make a change to a column. This chunk enables backwards compatibility because of a change on our end so that users wouldn't have to manually do it themselves for this special case.

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py:299]: Thanks for mentioning; those methods are what we're interfacing with in this method. Throughout the codebase we've been trying to use general functions where possible to simplify maintenance in the future, and I was trying to continue that here. My thinking was also that when we do the asset store refactor, part of the common store interface will be a method to send an API request to the respective stores. Does this approach make sense or am I overcomplicating things here?

File: tests/test_store.py
Comment (GiaJordan) [reply tests/test_store.py:221]: @thomasyu888 thanks for pointing that out! before this PR we weren't able to easily test `_update_table_uuid_column` due to the difficulty in triggering specific case where it's actually needed and the operations it was doing. It might be easier now that we've switched to using the rest calls

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py:2289]: Would the value returned by `self.synStore.syn.store(new_col)` be the same as what we're currently getting now / does the `syn.store` operation actually store anything in this case or just create a column type object? Also, do you know if there's a performance difference between using the API and the client for this operation?

File: schematic/store/synapse.py
Comment (thomasyu888) [reply schematic/store/synapse.py:299]: Ah - my apologies!  I should have looked at the code closer, but I had mistaken `request` for the `requests` package and I thought you were doing raw `requests.GET/POST/...` calls over the client.

There are multiple ways to think about this.  Since the asset store for 'tables' specifically are things like postgres, synapse tables, Bigquery, naturally it'll be different API calls which might be harder to encapsulate.  In terms of file stores, that would be using something like fsspec (which we do have on our roadmap to build out fs-synapse), which would be pretty different from storing files into tables.

heres some pseudo code for OOP structure.  Some of this could be abstracted in `schematic-db`, so this might be unnecessary.  I encourage everyone to learn about [SOLID principles](https://www.freecodecamp.org/news/solid-principles-explained-in-plain-english/)

```
from abc import ABC
from dataclasses import dataclass

@dataclass
class BaseAssetStore(ABC):
     connection_env: str
     ....
     @abstractmethod
     def store_table(self):
           ....

class SynapseStore(BaseAssetStore):
     def store_table():
           # your implementation here

@dataclass
class BaseTableUpdate(ABC):
    asset_store: BaseAssetStore,   # This is where you put in your store logic for different stores. Those could have different base classes and configurations
    tableToLoad: pd.DataFrame = None, 
    tableName: str = None, 
    datasetId: str = None, 
    existingTableId: str = None,
    restrict: bool = False
    
    @abstract_method
    def createTable():
         ...

@dataclass
class SynapseTableUpdate(BaseTableUpdate)
    asset_store: SynapseStore
    
    def createTable():
         ...
....

```

File: schematic/store/synapse.py
Comment (thomasyu888) [reply schematic/store/synapse.py:2289]: Yes, the value would be the same.  `syn.store` just creates a new column type object, basically the only difference here is you don't need to test all of this since it's tested in the client.

Behind the scenes, the client just does the API call.  If you're interested, `syn.store` will look for this function in any class: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/f3698af6c7d361c5a22d102d2625152933809ceb/synapseclient/table.py#L1517.  And it will call the `restPOST` command.  

You are essentially duplicating code and also formatting your own json when there's a Python object that you can use.

File: schematic/store/synapse.py
Comment (thomasyu888) [reply schematic/store/synapse.py:2163]: I definitely find this helpful!  That makes sense to me - Synapse Tables is definitely not a database.  In terms of backwards compatibility - I leave it to you all to decide whats best.  What I look for code is

1. Reusability
2. Readabilty
3. Testability

I challenge everyone to think broader.  Yes, this particular code may not be re-used here, but table column renames is probably something that is helpful to the Sage community.  By pulling that core functionality out, it would be easier to test, and you could take that directly and create a PR into the python client.

File: schematic/store/synapse.py
Comment (andrewelamb) [reply schematic/store/synapse.py:299]: I saw my name mentioned!

I think the eventual asset store refactor will do exactly this @thomasyu888, like schematic db does with [database backends](https://github.com/Sage-Bionetworks/schematic_db/blob/develop/schematic_db/rdb/rdb.py#L39)

**Comment by GiaJordan** [comment]
@thomasyu888 thank you so much for your comments, we're going to create tickets for these so that we can further optimize these portions of code in the future. From talking with @milen-sage during our 1 on 1 today we're going to implement your comments in coming PRs so that we can still deliver this functionality before I'm out next week.

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py:2289]: Ahh I see, that makes sense. I think in the future we'll switch to using the client for this operation and the api call for the renaming

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py:299]: That's a good point, the calls for other stores will likely be different than what we're using for synapse currently.
@linglp here's a heads up that this will likely change in the future, including a switch to just using the client functions directly at some point

File: schematic/store/synapse.py
Comment (thomasyu888) [inline schematic/store/synapse.py:2315]: I didn't realize this, but I dug deeper and I actually already have a function that does this: `syn._async_table_update` .  You can specify `wait=False` to speed up your code if you don't want to wait for the update to happen.


```suggestion
                   syn._async_table_update("syn51659783", changes=[columnChangeDict])
```

File: schematic/store/synapse.py
Comment (linglp) [reply schematic/store/synapse.py:299]: Thanks for the example @thomasyu888. I really like the idea of separating asset store base class and table operation. I think it makes sure that each class focuses on one thing. Thanks @GiaJordan for the reminder. I will keep in mind that the structure of the code might be changed in the future.

File: schematic/store/synapse.py
Comment (linglp) [reply schematic/store/synapse.py:2289]: @thomasyu888 for the example you gave, should the code be something like: 
```
from synapseclient.table import Column
new_col = Column(name='Id', columnType='STRING', maximumSize=64, maximumListLength=1) 
newColResponse = syn.store(new_col)
```

@GiaJordan if using the client directly is simplier, could you make the change in this PR and switch back to use the client? That will be easier to maintain in the future.

Comment (linglp) [reply schematic/store/synapse.py:2289]: I also dig into the client code more, and just like @thomasyu888 described, I think here is a duplicate of what the client is doing here: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/f3698af6c7d361c5a22d102d2625152933809ceb/synapseclient/client.py#L1035

When storing a column, `restPOST` is taking: `obj.postURI()` which is `/column`, and `obj.json` is the json object: 
```
{ "columnType": "STRING", "concreteType": "org.sagebionetworks.repo.model.table.ColumnModel", "id": "192889", "maximumListLength": 1, "maximumSize": 64, "name": "Id" }
```
which is like what you are doing with the `send_api_request` function

File: schematic/store/synapse.py
Comment (linglp) [reply schematic/store/synapse.py:2310]: @thomasyu888 I also noticed that there's a function called `_async_table_update` in client. Is that function also doing something similar? I saw: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/f3698af6c7d361c5a22d102d2625152933809ceb/synapseclient/client.py#L3439

Comment (thomasyu888) [reply schematic/store/synapse.py:2310]: See a comment I made below with this suggestion.

File: schematic/store/synapse.py
Comment (thomasyu888) [reply schematic/store/synapse.py:2289]: Yup, that's right!

File: tests/test_store.py
Comment (linglp) [inline tests/test_store.py:221]: @GiaJordan since I have merged my PR into your PR, this test will need to be updated?

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py:301]: @linglp we can remove

File: tests/test_store.py
Comment (GiaJordan) [reply tests/test_store.py:221]: We'll remove it along with `send_api_request`

Comment (GiaJordan) [reply tests/test_store.py]: that change has been pushed now

File: schematic/store/synapse.py
Comment (linglp) [inline schematic/store/synapse.py:2187]: @linglp @GiaJordan let me remove this before merge

Comment (linglp) [reply schematic/store/synapse.py]: @GiaJordan I just removed this line.

**Review by linglp — APPROVED** [review]

File: schematic/store/synapse.py
Comment (GiaJordan) [reply schematic/store/synapse.py]: Thank you!

**Pull request merged**
